File: frontend/src/lib/api/services/analytics.ts
import { apiClient } from "../client";
import type { GeoData, TimeSeriesData, Topic, ComparisonResult } from "@/types";

export interface DateRangeParams {
  date_from?: string;
  date_to?: string;
}

export interface HotHoursParams extends DateRangeParams {
  /** IANA timezone, e.g. "Asia/Almaty". Defaults to UTC server-side. */
  timezone?: string;
}

/** Single (day-of-week, hour) bucket of mentions in the caller's TZ. */
export interface HotHoursCell {
  day: number; // 0 = Sunday (Postgres `dow`)
  hour: number; // 0..23 in selected TZ
  mentions: number;
  reach: number;
  avg_sentiment: number;
  positive: number;
  neutral: number;
  negative: number;
}

export interface HotHoursAggregate {
  timezone: string;
  total_mentions: number;
  cells: HotHoursCell[];
  peak: HotHoursCell | null;
  top_cells: HotHoursCell[];
  active_days: { day: number; mentions: number }[];
  quietest_cells: HotHoursCell[];
}

export interface SourceBreakdownItem {
  source_id: string;
  name: string;
  type: string;
  base_url: string;
  icon: string | null;
  country: string | null;
  language: string | null;
  mentions_count: number;
  reach: number;
  share_pct: number;
  avg_sentiment: number;
  last_published_at: string | null;
}

export interface KeywordItem {
  word: string;
  count: number;
  is_hashtag: boolean;
  change_pct: number;
}

export interface TopLinkItem {
  url: string;
  domain: string;
  count: number;
  reach: number;
  last_seen: string | null;
}

export interface LanguageItem {
  language: string;
  name: string;
  count: number;
  share_pct: number;
}

export interface AnomalyEvent {
  date: string;
  mentions: number;
  z_score: number;
  type: "spike" | "drop";
}

export const analyticsApi = {
  async getGeoData(projectId: string, params: DateRangeParams = {}): Promise<GeoData[]> {
    const response = await apiClient.get<GeoData[]>(
      `/projects/${projectId}/analytics/geo`,
      { params }
    );
    return response.data;
  },

  async getHotHours(
    projectId: string,
    params: HotHoursParams = {}
  ): Promise<HotHoursAggregate> {
    const response = await apiClient.get<HotHoursAggregate>(
      `/projects/${projectId}/analytics/hot-hours`,
      { params }
    );
    return response.data;
  },

  async getEmotions(projectId: string, params: DateRangeParams = {}): Promise<{ averages: Record<string, number>; total_analyzed: number }> {
    const response = await apiClient.get(
      `/projects/${projectId}/analytics/emotions`,
      { params }
    );
    return response.data;
  },

  async getTopics(projectId: string, params: DateRangeParams = {}): Promise<Topic[]> {
    const response = await apiClient.get<Topic[]>(
      `/projects/${projectId}/analytics/topics`,
      { params }
    );
    return response.data;
  },

  async getTimeSeries(
    projectId: string,
    days = 30,
    params: DateRangeParams = {}
  ): Promise<TimeSeriesData[]> {
    const response = await apiClient.get<TimeSeriesData[]>(
      `/projects/${projectId}/analytics/time-series`,
      { params: { days, ...params } }
    );
    return response.data;
  },

  async getAnomalies(
    projectId: string,
    days = 30,
    z_threshold = 2.0
  ): Promise<AnomalyEvent[]> {
    const response = await apiClient.get<AnomalyEvent[]>(
      `/projects/${projectId}/analytics/events`,
      { params: { days, z_threshold } }
    );
    return response.data;
  },

  async getSourcesBreakdown(
    projectId: string,
    params: DateRangeParams & { limit?: number } = {}
  ): Promise<SourceBreakdownItem[]> {
    const response = await apiClient.get<SourceBreakdownItem[]>(
      `/projects/${projectId}/analytics/sources`,
      { params }
    );
    return response.data;
  },

  async getKeywords(
    projectId: string,
    params: DateRangeParams & { limit?: number } = {}
  ): Promise<KeywordItem[]> {
    const response = await apiClient.get<KeywordItem[]>(
      `/projects/${projectId}/analytics/keywords`,
      { params }
    );
    return response.data;
  },

  async getTopLinks(
    projectId: string,
    params: DateRangeParams & { limit?: number } = {}
  ): Promise<TopLinkItem[]> {
    const response = await apiClient.get<TopLinkItem[]>(
      `/projects/${projectId}/analytics/top-links`,
      { params }
    );
    return response.data;
  },

  async getLanguages(
    projectId: string,
    params: DateRangeParams = {}
  ): Promise<LanguageItem[]> {
    const response = await apiClient.get<LanguageItem[]>(
      `/projects/${projectId}/analytics/languages`,
      { params }
    );
    return response.data;
  },

  async compare(projectId: string, data: { type: string; item_ids: string[]; date_from?: string; date_to?: string }): Promise<ComparisonResult> {
    const response = await apiClient.post<ComparisonResult>(`/projects/${projectId}/compare`, data);
    return response.data;
  },
};

File: backend/app/api/routes/analytics.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.api.deps import get_current_user
from app.models.user import User
from app.schemas.analytics import ComparisonRequest
from app.services import analytics_service, emotion_service
from app.services.project_service import get_project

router = APIRouter()


@router.get("/projects/{project_id}/analytics/geo")
async def geo_data(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_geo_data(
        db, project_id, date_from=date_from, date_to=date_to
    )


@router.get("/projects/{project_id}/analytics/hot-hours")
async def hot_hours(
    project_id: str,
    timezone: str = Query("UTC", description="IANA timezone, e.g. Asia/Almaty"),
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Per-(day-of-week, hour) heatmap data with reach + sentiment,
    bucketed in the requested IANA timezone (defaults to UTC)."""
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_hot_hours(
        db,
        project_id,
        timezone=timezone,
        date_from=date_from,
        date_to=date_to,
    )


@router.get("/projects/{project_id}/analytics/emotions")
async def emotions(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Aggregated 8-emotion (Plutchik) distribution for the project,
    optionally constrained by ``date_from`` / ``date_to``.

    Returns averages, daily timeline and top mentions per emotion in a
    single payload — see ``emotion_service.aggregate_emotions``.
    """
    await get_project(db, project_id, current_user.id)
    return await emotion_service.aggregate_emotions(
        db, project_id, date_from=date_from, date_to=date_to
    )


@router.post("/projects/{project_id}/emotions/backfill")
async def backfill_emotions(
    project_id: str,
    limit: int | None = Query(None, ge=1, le=10_000),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Re-classify mentions whose ``mention.emotions`` is empty / NULL /
    legacy rule-based zeros. Trigger from the UI; the work is done
    inline so the user sees immediate progress in the toast.

    ``limit`` lets the operator dry-run on a small slice before
    committing tokens to the entire backlog.
    """
    await get_project(db, project_id, current_user.id)
    return await emotion_service.backfill_project(db, project_id, limit=limit)


@router.get("/projects/{project_id}/analytics/topics")
async def topics(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_topics_data(
        db, project_id, date_from=date_from, date_to=date_to
    )


@router.get("/projects/{project_id}/analytics/sources")
async def sources_breakdown(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_sources_breakdown(
        db, project_id, date_from=date_from, date_to=date_to, limit=limit
    )


@router.get("/projects/{project_id}/analytics/keywords")
async def keywords(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    limit: int = Query(50, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_keywords(
        db, project_id, date_from=date_from, date_to=date_to, limit=limit
    )


@router.get("/projects/{project_id}/analytics/top-links")
async def top_links(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    limit: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_top_links(
        db, project_id, date_from=date_from, date_to=date_to, limit=limit
    )


@router.get("/projects/{project_id}/analytics/languages")
async def languages_breakdown(
    project_id: str,
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_languages_breakdown(
        db, project_id, date_from=date_from, date_to=date_to
    )


@router.get("/projects/{project_id}/analytics/time-series")
async def time_series(
    project_id: str,
    days: int = Query(30, ge=1, le=365),
    date_from: str | None = None,
    date_to: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_time_series(
        db, project_id, days=days, date_from=date_from, date_to=date_to
    )


@router.get("/projects/{project_id}/analytics/events")
async def anomaly_events(
    project_id: str,
    days: int = Query(30, ge=7, le=365),
    z_threshold: float = Query(2.0, ge=1.0, le=5.0),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return await analytics_service.get_anomaly_events(
        db, project_id, days=days, z_threshold=z_threshold
    )


@router.post("/projects/{project_id}/compare")
async def compare(
    project_id: str,
    data: ComparisonRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)

    # Security: only let the user compare projects they actually own.
    # Silently drop foreign IDs instead of 403'ing so the UI stays useful
    # if someone navigates with a stale URL.
    from app.models.project import Project as _Project
    from sqlalchemy import select as _select

    owned_ids: set[str] = set()
    if data.item_ids:
        rows = (
            await db.execute(
                _select(_Project.id).where(
                    _Project.id.in_(data.item_ids),
                    _Project.owner_id == current_user.id,
                )
            )
        ).all()
        owned_ids = {row.id for row in rows}

    safe_ids = [i for i in data.item_ids if i in owned_ids]
    if not safe_ids:
        # Always include the source project at minimum so the response is non-empty.
        safe_ids = [project_id]

    return await analytics_service.get_comparison(
        db, project_id, data.type, safe_ids, data.date_from, data.date_to
    )


@router.get("/projects/{project_id}/compare")
async def compare_get(
    project_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await get_project(db, project_id, current_user.id)
    return {"type": "projects", "items": [], "metrics": []}
